### PR TITLE
Implement the key related apis

### PR DESCRIPTION
**Pull request by pengisgood**
Refer to the document [here](https://docs.ipfs.io/reference/http/api/#api-v0-key-gen)

### CODE REVIEW

**Comment by pengisgood** [comment]
@Stebalien  Please have a review. :)

**Comment by pengisgood** [comment]
@aschmahmann Is there any progress for the review?

**Comment by aschmahmann** [comment]
Hey, @pengisgood. I didn't have time to look at things things last week, but should be able to take a look this week. Thanks for the PR.

**Comment by pengisgood** [comment]
ping @aschmahmann

**Comment by aschmahmann** [comment]
Sorry... I haven't forgotten about this just gotten a bit busy. Will try to take a look today, or early next week.

File: key_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+	"github.com/cheekybits/is"
+)
+
+func TestKeyGen(t *testing.T) {
+	is := is.New(t)
+	s := NewShell(shellUrl)
+
+	key, err := s.KeyGen(context.Background(), "testKey", KeyGen.Type("ed25519"), KeyGen.Size(2048))
+	is.Nil(err)
```
Comment (aschmahmann) [inline key_test.go:15]: While this passes when run against go-ipfs it's essentially a bug. ed25519 implicitly has a size associated with it (and that size is 256). Instead we should test:

1. ed25519 with no size works
2. rsa with no size works
3. rsa with a valid custom size (e.g. 4096) works
4. rsa with an invalid custom size (e.g. 1024, which is too small) fails

File: key_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+func TestKeyList(t *testing.T) {
+	is := is.New(t)
+	s := NewShell(shellUrl)
+
+	keys, err := s.KeyList(context.Background())
+	is.Nil(err)
+
+	is.Equal(len(keys), 1)
+	is.Equal(keys[0].Name, "self")
+	is.NotNil(keys[0].Id)
```
Comment (aschmahmann) [inline key_test.go:33]: We should verify the KeyList works with multiple return values. Try inserting a few keys and then doing a KeyList (then remove them so we leave the state as we found it).

File: key_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+func TestKeyRename(t *testing.T) {
+	is := is.New(t)
+	s := NewShell(shellUrl)
+
+	key, err := s.KeyGen(context.Background(), "test1")
+	is.Nil(err)
+
+	out, err := s.KeyRename(context.Background(), "test1", "test2", false)
+	is.Nil(err)
```
Comment (aschmahmann) [inline key_test.go:44]: We should also verify that we properly handle errors, such as if we rename a non-existent key.

File: key_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+func TestKeyRm(t *testing.T) {
+	is := is.New(t)
+	s := NewShell(shellUrl)
+
+	key, err := s.KeyGen(context.Background(), "testKey")
+	is.Nil(err)
+
+	keys, err := s.KeyRm(context.Background(), "testKey")
+	is.Nil(err)
```
Comment (aschmahmann) [inline key_test.go:63]: We should also verify that we properly handle errors, such as if we remove a non-existent key.

**Comment by pengisgood** [comment]
> @pengisgood this looks pretty good, thanks for the PR. I added some comments about a few extra tests it would be great to have before merging this.
> 
> Ping me when they're done and we'll get this merged 🚀

Sure, I'll add more test cases tomorrow. Thanks.

**Comment by pengisgood** [comment]
@aschmahmann I added more test cases.

**Comment by pengisgood** [comment]
ping @aschmahmann

**Comment by pengisgood** [comment]
This PR takes almost 3 weeks now. Is it possible to speed up a little bit to get it merged? Honestly, I'm a little frustrated. I just want to give a rush to get small things done.

File: key_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+func TestKeyGen(t *testing.T) {
+	is := is.New(t)
+	s := NewShell(shellUrl)
+
+	defer func() {
+		_, err := s.KeyRm(context.Background(), "testKey1")
+		is.Nil(err)
+	}()
+	key1, err := s.KeyGen(context.Background(), "testKey1", KeyGen.Type("ed25519"), KeyGen.Size(2048))
```
Comment (aschmahmann) [inline key_test.go:18]:
```suggestion
	key1, err := s.KeyGen(context.Background(), "testKey1", KeyGen.Type("ed25519"))
```

per https://github.com/ipfs/go-ipfs-api/pull/220#discussion_r450294770 having ed25519 + 2048 bits does not make sense. Could you please remove the Size option here.

**Comment by pengisgood** [comment]
@aschmahmann changed as you suggested.

**Pull request merged**